Add Weights and Biases Logger

drozzy opened the pull request:
This fixes #426.

# Summary of changes

1. I renamed BasicLogger to TensorboardLogger and updated references everywhere.
2. I did NOT change the logging api, since I didn't want to mess stuff up. I am not sure what the right signature for write should be. I just wanted to make it work and not break stuff.
3. I fixed the imports/exports. Double check the `utils/__init__.py` please.
4. I added wandb as a dependency in setup. Perhaps it can only be added as a test-time dependency?
5. I ran all the pep/type/unit tests. All passed except one failed test (I'm not sure if it was failing before, but I can't see how I would have broken it):
```
test/modelbased/test_psrl.py::test_psrl reward threshold: 3400
FAILED
FAILED test/modelbased/test_psrl.py::test_psrl - KeyError: 'rew'

```
6. I get these errors for docstrings and I've no idea how to fix them:
```
tianshou/utils/logger/wandb.py:8 in public class `WandBLogger`:
        D208: Docstring is over-indented
tianshou/utils/logger/wandb.py:8 in public class `WandBLogger`:
        D209: Multi-line docstring closing quotes should be on a separate line
```
7. All other doc tests pass.

P.S.: As a side note, the unit tests take a really long time to run.

Trinkle23897 commented:
> I renamed BasicLogger to TensorboardLogger and updated references everywhere.

better to keep BasicLogger for compatibility, I'll fix that

drozzy commented:
> > I renamed BasicLogger to TensorboardLogger and updated references everywhere.
> 
> better to keep BasicLogger for compatibility, I'll fix that

I find BaseLogger BasicLogger naming confusing a bit. But maybe we can do it in another release or something.

Trinkle23897 commented:
I know, I mean
```
class BasicLogger(TensorboardLogger):
  def __init__(self, *args, **kwargs):
    warnings.warn("better to use TensorboardLogger")
    super().__init__(*args, **kwargs)
```

drozzy commented:
Oooh i see, good idea!

Also, for this I usually see people put something like "Deprecated: use TensorboardLogger instead"

Trinkle23897 commented:
```python
In [1]: from tianshou.utils import BasicLogger

In [2]: a=BasicLogger(None)
/home/trinkle/github/tianshou-new/tianshou/utils/logger/tensorboard.py:135: UserWarning: Deprecated soon: BasicLogger has renamed to TensorboardLogger in #427.
  warnings.warn(
```

Trinkle23897 commented:
> 2\. I did NOT change the logging api, since I didn't want to mess stuff up. I am not sure what the right signature for write should be. I just wanted to make it work and not break stuff.

I'm refactoring this part now.

Trinkle23897 commented on docs/tutorials/cheatsheet.rst:
```diff
@@ -40,8 +40,8 @@ This is related to `Issue 349 <https://github.com/thu-ml/tianshou/issues/349>`_.
-2. Use ``BasicLogger`` which contains a tensorboard;
-3. To adjust the save frequency, specify ``save_interval`` when initializing BasicLogger.
+2. Use ``TensorboardLogger``;
+3. To adjust the save frequency, specify ``save_interval`` when initializing TensorboardLogger.
```
Is it possible to resume an experiment with wandb? like #350

drozzy replied:
I think so, but I've never dealt with it before. I suggest waiting for people who have that specific need to raise an issue. 
Because I think to resume it you need to start it differently - and it may be extra complexity not everyone needs.

Trinkle23897 commented on tianshou/policy/modelfree/ppo.py:
```diff
@@ -22,7 +22,7 @@ class PPOPolicy(A2CPolicy):
         where c > 1 is a constant indicating the lower bound.
         Default to 5.0 (set None if you do not want to use it).
-    :param bool value_clip: a parameter mentioned in arXiv:1811.02553 Sec. 4.1.
+    :param bool value_clip: a parameter mentioned in arXiv:1811.02553v3 Sec. 4.1.
```
#429

Trinkle23897 commented:
Should I add wandb logger test such as https://wandb.ai/tianshou/tianshou-new-test_continuous/runs/3gcpla04?workspace=user-tianshou?

drozzy commented:
Those runs are not visible to me.
Regarding tests - it's up to you. I'm not really sure how you could test it.

Trinkle23897 commented:
Okay so plz review my change

drozzy commented on tianshou/utils/logger/wandb.py:
```diff
@@ -0,0 +1,44 @@
+    ) -> None:
+        super().__init__(train_interval, test_interval, update_interval)
+
+    def write(self, step_type: str, step: int, data: LOG_DATA_TYPE) -> None:
```
Oh I see how you did this. Cool.

drozzy replied:
Did you test it? Or do you want me to test it?

drozzy commented:
Looks good to me, I can start using to make sure everything works.
